### ISSUE TITLE
Training with flash-attention causing loss to differ

### ISSUE THREAD

**Issue by jiangzzsss** [Dao-AILab/flash-attention#109]
Hi, since flash attention can reduce time cost and device memory usage significantly, we'd like to use it in our model training.
However, we've encountered a loss difference with baseline when training:
1. we use it in inference, it works just fine
2. the same code in training infects our model's performance.

Note: our model is a diffusion model with slight difference and trained with fp32. To fit flash-attention, we change the related part in CrossAttention to fp16 and then change the result back to fp32.
We train with 8 gpus and ddp, inference only use 1 gpu.
### Environments Info:
- python 3.7.11
- pytorch 1.11+cu113
- lash-attn 0.2.2
- A100 80G

### partial code
```
        qb, qn, qd = q.shape
        kb, kn, kd = k.shape
        vb, vn, vd = v.shape
        using_flash = False
        if((qd // h) <= 128 and (kd // h) <= 128 and (vd // h) <= 128): using_flash = True
        if(opt_helper.using_flash and using_flash):
            # print('using flash')
            opt_helper.flash_cnt += 1
            q = q.reshape(qb * qn, h, qd//h).to(dtype=torch.float16)
            k = k.reshape(kb * kn, h, kd//h).to(dtype=torch.float16)
            v = v.reshape(vb * vn, h, vd//h).to(dtype=torch.float16)
            max_seqlen_q = qn
            max_seqlen_k = kn
            cu_seqlens_q = torch.arange(
                0,
                (qb + 1) * qn,
                step=qn,
                dtype=torch.int32,
                device=q.device,
            )
            cu_seqlens_k = torch.arange(
                0,
                (kb + 1) * kn,
                step=kn,
                dtype=torch.int32,
                device=k.device,
            )
            output = flash_attn_unpadded_func(
                q, k, v,
                cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k,
                dropout_p = 0.,
                softmax_scale = self.scale,
            )
            out = output.reshape(qb, qn, qd).float()
```
**In pytorch AMP, softmax using fp32, we kind of wonder would this be an issue? our query num(qn, kn) might be large(e.g. 4096)**

**Comment by tridao** [comment]
FlashAttention should give the same output as standard attention, up to some numerical tolerance (we have tests for this). Internally softmax is done in fp32 as well (matrix multiply Q @ K^T and attn @ V are done in fp16/bf16).

The comparison we usually make is:
(FlashAttention in fp16/bf16 - standard attention in fp32) vs (standard attention in fp16/bf16 - standard attention in fp32).

In all of our use cases the errors between the two are around the same.

For your case:
1. Can you print out these two numerical differences to compare?
2. Can you try running standard attention (softmax(Q @ K^T) @ V) in fp16 to see if that affects your model quality. If that's the case, that means that the model is simply sensitive to numerical differences and you'd want to run attention in fp32 anyway (so just use standard attention in that case).
For the use cases we've seen (training language models & vision transformers), swapping out standard attention in fp16/bf16 with FlashAttention in fp16/bf16 does not change the model quality.

**Comment by ranjiewwen** [comment]
@tridao thanks you work, for diffusion model generate image pixel, maybe sensitive to numerical differences. we test inference model, and result only 1e-2 can align between FlashAttention in fp16- standard attention in fp16, and generates poor image quality. any suggestion ？ Retrain can solve the problem ? but it too expensive。

**Comment by tridao** [comment]
What are the numerical errors
(FlashAttention in fp16 - standard attention in fp32) vs (standard attention in fp16 - standard attention in fp32)?

Are those numerical errors for the whole model or just one attention layer?

**Comment by ranjiewwen** [comment]
@tridao thanks for your reply, we test FlashAttention in fp16 vs standard attention in fp16,  numerical errors just one attention layer.

**Comment by tridao** [comment]
@ranjiewwen What are the numerical errors
(FlashAttention in fp16 - standard attention in fp32) vs (standard attention in fp16 - standard attention in fp32)?

**Comment by ranjiewwen** [comment]
@tridao (FlashAttention in fp16 - standard attention in fp16) > 1e-3

**Comment by tridao** [comment]
@ranjiewwen what's (FlashAttention in fp16 - standard attention in **fp32**)?
and (standard attention in fp16 - standard attention in **fp32**)?

The reference output should be standard attention in **fp32**.

**Comment by ranjiewwen** [comment]
@tridao i know, but when we only use flashattention for inference, the train model use pytorch fp16,  so we compare the train layer use pytorch fp16 - inference layer use flashattention fp16, the result diff > 1e-3.

**Comment by tridao** [comment]
Can you get (standard attention in fp16 - standard attention in fp32) anyway?
If this difference is > 1e-3, does that imply that running inference with standard attention in fp32 would also produce unacceptable results? If your application is that sensitive (running inference with **higher** precision produces unacceptable results) then you should not change anything between training and inference.

Depending on the magnitude of the values, a difference of 1e-3 is expected for fp16 precision. For example, doing any simple arithmetic on fp16 values already produces numerical error of 5e-4 or 1e-3:
```
In [1]: import torch

In [2]: a = torch.randn(100, dtype=torch.float16, device='cuda')

In [3]: (a + 0.3 - 0.3 - a).abs().max()
Out[3]: tensor(0.0010, device='cuda:0', dtype=torch.float16)
```

**Comment by ranjiewwen** [comment]
ok, thanks.

**Comment by piegu** [comment]
Hi @tridao,
I ran the [code of Philipp Schmid (HF) to finetune Llama 2 7B](https://www.philschmid.de/sagemaker-llama2-qlora) with and without Flash Attention 2. 

**The loss is higher WITH Flash Attention 2.** 
Do you have any idea why?

**Comment by tridao** [comment]
Loss should be around the same. If there's a difference, you should print out the output of standard attention and FlashAttention, and check that they are close numerically. That should tell you if you're using it correctly.

**Comment by piegu** [comment]
Thanks @tridao for your answer. Here the training code I use (full code at [run_qlora.py](https://github.com/philschmid/llm-sagemaker-sample/blob/main/scripts/run_qlora.py)).

This is the classic HF trainer code with the `use_flash_attention_2=True/False` argument with the latest official libraries versions.

My use case: I finetune the Llama 2 7B model on custom dataset on 2 GPUs RTX 3090 (24GB RAM). I use chunck of 4096 tokens (I need that because of the content size of my prompts).
- with "use_flash_attention_2=False": my training batch size is 1 (gradient_accumulations_steps = 2)
- with "use_flash_attention_2=True": my training batch size is 5 (gradient_accumulations_steps = 1)

What do you think? With this complementary information, is there a reason to get a higher loss with `use_flash_attention_2=True`?

```
def training_function(script_args, training_args):

    # load dataset
    dataset_train = load_from_disk(script_args.path_to_dataset_train)
    dataset_eval = load_from_disk(script_args.path_to_dataset_eval)
    
    # load model from the hub with a bnb config
    bnb_config = BitsAndBytesConfig(
        load_in_4bit=True,
        bnb_4bit_use_double_quant=True,
        bnb_4bit_quant_type="nf4",
        bnb_4bit_compute_dtype=torch.bfloat16,
    )

    model = AutoModelForCausalLM.from_pretrained(
        script_args.model_id,
        use_cache=False if training_args.gradient_checkpointing else True,  # this is needed for gradient checkpointing
        device_map="auto",
        use_flash_attention_2=script_args.use_flash_attn,
        quantization_config=bnb_config,
    )

    # create peft config
    model = create_peft_model(
        model, gradient_checkpointing=training_args.gradient_checkpointing, bf16=training_args.bf16
    )

    # Create Trainer instance
    trainer = Trainer(
        model=model,
        args=training_args,
        train_dataset=dataset_train,
        eval_dataset=dataset_eval,
        data_collator=default_data_collator,  # no special collator needed since we stacked the dataset
    )

    # Start training
    trainer.train()
```

**Comment by tridao** [comment]
Did you print out the output of standard attention and FlashAttention, and check that they are close numerically?

**Comment by piegu** [comment]
> Did you print out the output of standard attention and FlashAttention, and check that they are close numerically?

Until now, no. Can you give me a code for doing that? Thank you.

**Comment by tridao** [comment]
I can't help with that, sorry. You'd need to edit the model file (e.g. modeling_llama.py on HF if you're training llama).

**Comment by piegu** [comment]
The model file for training is here [run_qlora.py](https://github.com/philschmid/llm-sagemaker-sample/blob/main/scripts/run_qlora.py).

When I ask you some code to do what you asked for (ie, "print out the output of standard attention and FlashAttention"), it's because I'm not sure of what your are asking for. What means "output of standard attention and FlashAttention"?

**Comment by tridao** [comment]
something like:
```
output_reference = attention(query.float(), key.float(), value.float())
output_baseline =  attention(query, key, value)
output_flash = flash_attention(query, key, value)
print((output_baseline - output_reference).abs().max())
print((output_flash - output_reference).abs().max())
```

**Comment by piegu** [comment]
Thanks @tridao but I don't see how to implement this pseudocode.

In fact, I just tested another idea: instead of using a `Context Window` of 4096 (Llama 2), I divided it by 2 until 2048.

My first results are as follows:

`use_flash_attention_2=False`

- first evaluation
  - train loss = 0.7167
  - eval loss = 0.5650
- second evaluation
  - train loss = 0.4035
  - eval loss = 0.3455

`use_flash_attention_2=True`

- first evaluation
  - train loss = 0.7144
  - eval loss = 0.5628
- second evaluation
  - train loss = 0.4017
  - eval loss = 0.3452

This time, the losses are similar (the "use_flash_attention_2=True" ones are even smaller).

Do you think the `Context Window` might be the "problem" of my case?

**Comment by tridao** [comment]
Sorry I can't help debug your training. If results are correct at 2k I don't see a reason it wouldn't be correct at 4k.
If you want to debug, you should go to the modeling file, e.g. make a copy of `modeling_llama.py` from huggingface, and put breakpoint / print statements there.
